add full name calculation

gregbunce opened the pull request:
just getting things going, so i wanted to check in before i got too far.

gregbunce committed:
are we on the right track

steveoh commented on .vscode/settings.json:
```diff
@@ -43,5 +43,6 @@
     "yapf"
-  ]
+  ],
+  "python.pythonPath": "C:\\Users\\gbunce\\AppData\\Local\\ESRI\\conda\\envs\\arcgispro-py3-clone\\python.exe"
```
**nitpick**: this doesn't need to be set any longer with vs code. I thought it was out of experimentation so you might just need to update or opt into the experiment with 

```json
"python.experiments.optInto": [
    "DeprecatePythonPath - experiment"
]
```

to your global settings.json (ctrl + shift + p) `Preferences: Open Settings (JSON)`.

A lot of have been doing it without issues so it should be fine.

gregbunce replied:
gotcha. thanks for the info. 

so, it's removed from the project's .vscode/settings.json file and i added that line to my c;\\users\\....\\settings.json.  does that sound right?

steveoh replied:
Yup

steveoh commented on src/rules/common.py:
```diff
@@ -30,7 +30,7 @@
 
 CONVERT_NULL_TO_ZERO = '''
-write arcade here
+IIf(isempty($feature.{0}) || $feature.{0} == <null>, '0', true);
```
**question**: does this create a string 0 or a number 0 here?

gregbunce replied:
i think i see what you're saying... b/c of the single quotes?  maybe i should remove the quotes.  the field where the values will be stored is set to integer.

gregbunce replied:
removed the single quotes

steveoh commented on src/rules/roads.py:
```diff
@@ -28,20 +29,20 @@
-name_calculation = Calculation('Name', 'NAME', 'Upper(\'NAME\')')
-alias_name_calculation = Calculation('Alias Name', 'A1_NAME', 'Upper(\'A1_AME\')')
-alias_alternate_name_calculation = Calculation('Alternate Alias Name', 'A2_NAME', 'Upper(\'A2_NAME\')')
+# name_calculation = Calculation('Name', 'NAME', 'Upper(\'NAME\')')
```
**suggestion**: prefer removing the code rather than commenting it out

gregbunce replied:
gotcha.  will do.  just left it there as my next step.

steveoh reviewed (commented):
yeah this looks like you're on the right track

gregbunce committed:
get branch update from origin

gregbunce committed:
Merge branch 'feat/road-name-calculation' of https://github.com/agrc/utrans-attribute-rules into feat/road-name-calculation
i
realign local with remote

gregbunce committed:
remove hardcoded path

gregbunce commented:
i made a few changes based on your suggestions.  should i mark it as 'ready for review'?

steveoh commented:
Can you send me or better yet commit the xml schema doc so I can load in the db and test this?

gregbunce committed:
add xml schema doc

gregbunce commented:
let me know if you need some data as well.  i exported the 'schema only'
![image](https://user-images.githubusercontent.com/16906316/103025863-e5f81980-450f-11eb-88cb-31f032d3e853.png)

gregbunce commented:
also, the xml schema contains all the needed layers (data store) for the attribute rules.  let me know if you only want the roads layer (the layer we're putting the attribute rules on).  
![image](https://user-images.githubusercontent.com/16906316/103026049-49824700-4510-11eb-8163-cbca4738f7ee.png)

steveoh commented:
> also, the xml schema contains all the needed layers (data store) for the attribute rules. l

The schema imported but when I tried to append data to some of the datasets so the rules will function the schema's didn't match and it failed. I ended up having to delete and then copy in the data. Let's do a schema export with data to eliminate these steps.

steveoh committed:
docs: update reference data

steveoh committed:
fix: make rules execute

steveoh committed:
chore: create version information table

steveoh committed:
fix: editable is not yes no any longer

steveoh committed:
refactor: remove for py3

steveoh committed:
fix: syntax error

steveoh committed:
fix: return null instead of setting value to true

this might not be the expected behavior

steveoh committed:
chore: remove readme

steveoh committed:
fix: add localhost connection

steveoh committed:
chore: add proproject

steveoh committed:
chore: ignore pro project cruft

steveoh committed:
chore: update dictionary

steveoh reviewed (changes requested):
This seems to be in a draft state still but going in the right direction. 

There are quite a few issues that are frustrating for a reviewer and I will list them so we can have more productive reviews in the future.

1. The pr text is a bit light. It says you're hoping to be going in the right direction but don't really state the known issues or things you are working out.
2. You don't mention what the full name calculation is expected to do. We don't have tests yet so some text on expectations would be very helpful.
3. The `ar.py` wasn't able to run and add the rules for testing.
4. The pro project was missing with the connection sde file.
5. The `roads_edit` fc was not able to have attribute rules added to it since it is missing global ids.

Otherwise yes it looks like you aren't in the weeds making timely mistakes.

gregbunce commented:
thanks for the useful feedback and for the work you put into this pr to get it up and running.  all good points - and noted on my end.

I'll pull this branch to my local repo.   then...

1. add globalid to `roads_edit` 
2. replace the existing XML schema with one containing the data.
3. test your commits
4. push my local branch back to the remote
5. 'ready for review'

Does that sound right?

thanks again.  once i get this set up, i feel as though I'll be able to make more timely progress on the project - doing local testing and a more focused pr. 

should i do my local testing using the pro project you added to the repo?

steveoh commented:
You bet. This branch should already exist in your local repo unless you did all the work in the github.com ui which I doubt. So you can most likely just checkout feat/road-name-calculation if you're not already on it and git pull.

1. yes
2. yes
3. yes and make the necessary changes to make sure the rules are creating the proper results
4. the commits you make to get the rules properly functioning, yes, you will then push to this existing branch. no need to make any new branches
5. yes - when everything is working as expected

Yes we should align on that pro project. There might be some broken items, i'm thinking paths etc but we should work that out so we are working on identical items.

gregbunce commented:
k, perfect.  yup, i pulled your changes to my local branch.  

i'm trying to work out the pro project now.  working on how to use the shared sde connection file.  is that possible?  it looks like you used os auth.  

is there a way we can use a common Instance name or should i just use my local connection string and just work on sharing the pro project?  (if that makes sense)

here's how i see your connection file:
![image](https://user-images.githubusercontent.com/16906316/103371847-b5732b00-4a8d-11eb-9e92-b01ecfda3c98.png)

steveoh commented:
let's try to go with the local style as that will work for anyone with a local sql server and won't have any username/passwords in it.

gregbunce committed:
chore: update readme

gregbunce committed:
chore: add road xml data

gregbunce committed:
chore: make reference data a list

gregbunce committed:
update sde file name

gregbunce committed:
update stage and prod sde conn file names

gregbunce committed:
fix null to zero expression

this is now working in arcgis pro testing

gregbunce committed:
fix null to zero expression

this is now working in arcgis pro testing

gregbunce committed:
fix upper calculations

and set triggers to update; also rename ar rules to reflect the associated field name

gregbunce commented:
all of the calculation rules appear to be working in arcgis pro during my manual testing.  the next step is to now write some testing in test_roads.py.  i put some time aside next week to do so (beginning Monday).  

also, currently, we have one constrain rule that we've created (predir_domain_constraint) but there are no domains in the db.  how should we handle that?  i guess they didn't carry over in the XML schema.  i can export the existing domains in utrans (the 'real' utrans) to tables and then we can add them to the local utrans db.  does that work?

gregbunce committed:
update arcpy install to conda

gregbunce committed:
add rules testing for fullname

steveoh commented on pro-project/.pyHistory:
```diff
@@ -0,0 +1,18 @@
+﻿table = "Roads_Edit"
```
**chore**: add this file to the .gitignore please

steveoh commented on readme.md:
````diff
@@ -47,7 +47,7 @@
 
     ```sh
-    conda create --clone arcgispro-py3 --name utrans
+    conda create --name utrans
````
**praise**: niiiiiice!

steveoh commented on tests/test_roads.py:
```diff
@@ -7,32 +7,25 @@
-
 table_name = 'Roads_Edit'
-sde = os.path.join(os.path.dirname(__file__), '..', 'pro-project', 'localhost.sde')
+sde = os.path.join(os.path.dirname(__file__), '..', 'pro-project', 'localhost@utrans.sde')
 TABLE = os.path.join(sde, table_name)
```
**nitpick**: we should switch this to use the newer [pathlib](https://docs.python.org/3/library/pathlib.html) which is the recommended library for the future.

something like the following should suffice
```suggestion
from pathlib import Path

sde = Path(__file__).parent / '..' / 'pro-project', 'localhost@utrans.sde'
TABLE = str(sde / table_name)
```

steveoh commented on tests/test_roads.py:
```diff
@@ -7,32 +7,25 @@
-    with arcpy.da.UpdateCursor(TABLE, ['OID@'], where_clause="FacilityName IN ('0', '1')") as cursor:
-        for _ in cursor:
-            cursor.deleteRow()
+    #arcpy.management.TruncateTable(TABLE)
```
**chore**: we need to figure out the reset issue otherwise the subsequent tests will fail

steveoh reviewed (commented):
looks like you're on the right track!
